nodes validate payments too

joshuef opened the pull request:
…te validation## Description

reviewpad:summary

maqi commented on sn_networking/src/lib.rs:
```diff
@@ -621,6 +622,43 @@ impl Network {
             .await)
     }
 
+    pub async fn get_store_cost_from_network(&self, dbc_id_key: NetworkAddress) -> Result<Token> {
```
this shall be a more generic record_entry, i.e. the target could be a chunk, register or DBC itself ?

maqi commented on sn_networking/src/lib.rs:
```diff
@@ -621,6 +622,43 @@ impl Network {
+            .into_iter()
+            .collect_vec();
+
+        let request = Request::Query(Query::GetStoreCost(dbc_id_key));
```
we already sending the Query to the specific peers.
Hence the `address` of the target is not that matter, instead the `size` will be more interested?

joshuef replied:
I'm not sure what you mean here

maqi replied:
I mean for the `query` request, we already `get_clostest_peers` within the above step according to the `dbc_id_key` address, and then send the request directly to peers one by one.
Hence, the `address content` with the `Query::GetStoreCost` doesn't matter anymore. Unless the cost is related to the data size to be uploaded, then what shall be wrapped in the Query shall be `data size` instead of `data address`.

joshuef replied:
Ah right. Yeh we can drop that entirely I think. the price is per record, so size doesn't matter to nodes, client can work it out.

joshuef replied:
Alllthough, if we drop the address, we don't have `Query.dst()`, so things get complicated. I'd leave that I think.

maqi commented on sn_networking/src/lib.rs:
```diff
@@ -621,6 +622,43 @@ impl Network {
+
+        // loop over responses, generating an avergae fee and storing all responses along side
+        let mut all_costs = vec![];
+        for response in responses.into_iter().flatten() {
```
in case there is error among the responses, say due to time_out, then the `flatten` will just swallow the error right?
in that case, shall we ensure the final OK responses still meet a minium len, say more than 3 ?

joshuef replied:
That's handled in the `get_fee_from_storecost_quotes`

maqi commented on sn_node/src/put_validation.rs:
```diff
@@ -89,6 +91,7 @@ impl Node {
     pub(crate) async fn validate_and_store_chunk(
         &self,
         chunk_with_payment: ChunkWithPayment,
+        validate_payment: bool,
```
as mentioned at the discussion https://maidsafe.discourse.team/t/pricing-data-simply/808/13
For ReplicatedData, we may still need to validate the payment is valid, just the `amount` doesn't matter
For original `PutRecord`, we not only need to validate the payment is valid, but also need to confirm the amount is correct

joshuef replied:
Updated for this 👍

joshuef committed:
chore(node): update put validations for flexible storage cost

joshuef committed:
feat(node): validate payments on kad:put

BREAKING CHANGE: Adds in node side validation of network messaging
around store cost

joshuef committed:
chore(node): always validate chunks, make erroring out for payment optional

joshuef committed:
chore(networking): update arg naming

joshuef committed:
Merge branch 'main' into Feat-ValidatePaymentsAtNodeToo
